### PR TITLE
IODEMO: extend server statistics

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
```
[1619535443.311214] [DEMO] read:90184 ops, write:90162 ops, active connections:3, min:57176/223.344MB, (2.1.3.9:58507), max:61829/241.52MB, (2.1.3.9:55453), buffers:3, pin bytes:13594624 regions:10 evict:0
[1619535444.312027] [DEMO] read:92721 ops, write:92726 ops, active connections:3, min:59457/232.254MB, (2.1.3.9:58507), max:123881/483.91MB, (2.1.3.9:51737), buffers:3, pin bytes:13594624 regions:10 evict:0
[1619535445.312142] [DEMO] read:93596 ops, write:93577 ops, active connections:3, min:60224/235.25MB, (2.1.3.9:58507), max:64174/250.68MB, (2.1.3.9:55453), buffers:3, pin bytes:13594624 regions:10 evict:0
```
@yosefe @alinask pls review

### CODE REVIEW

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -609,6 +609,44 @@ class DemoServer : public P2pDemoCommon {
...
+
+        long get_total_count() const {
+            return _read_req_count + _write_req_count;
+        }
+
+        long get_total_bytes() const {
+            return _read_req_bytes + _write_req_bytes;
+        }
+
+        void read_add(long bytes) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: maybe "read_started(size_t size)" or "read_done(size_t size)"

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -680,6 +718,11 @@ class DemoServer : public P2pDemoCommon {
         iov->init(msg->data_size, _data_chunks_pool, msg->sn, opts().validate);
         cb->init(iov, &_curr_state.read_count);
 
+        std::map<uint32_t, conn_state_t>::iterator it =
+                _conn_stat_map.find(conn->id());
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: shall we check it doesn't hurt perf?

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -697,17 +740,28 @@ class DemoServer : public P2pDemoCommon {
...
+        assert(it != _conn_stat_map.end());
+        it->second.write_add(msg->data_size);
         recv_data(conn, *iov, msg->sn, w);
     }
 
     virtual void dispatch_connection_accepted(UcxConnection* conn) {
+        bool res = _conn_stat_map.insert(std::make_pair(conn->id(),
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: maybe _conn_stat_map[conn->id] = conn_state_t() ?

File: test/apps/iodemo/io_demo.cc [hunk 4]
```diff
@@ -697,17 +740,28 @@ class DemoServer : public P2pDemoCommon {
...
     }
 
     virtual void dispatch_connection_error(UcxConnection *conn) {
         LOG << "deleting connection with status "
             << ucs_status_string(conn->ucx_status());
         assert(conn->is_established());
+        std::map<uint32_t, conn_state_t>::iterator it =
+                _conn_stat_map.find(conn->id());
+        assert(it != _conn_stat_map.end());
+        _conn_stat_map.erase(it);
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:761]: can we erase by key?

File: test/apps/iodemo/io_demo.cc [hunk 5]
```diff
@@ -743,12 +797,46 @@ class DemoServer : public P2pDemoCommon {
         memory_pin_stats_t pin_stats;
         memory_pin_stats(&pin_stats);
 
-        LOG << "read:" << _curr_state.read_count - _prev_state.read_count
+        std::map<uint32_t, conn_state_t>::iterator it_min, it_max, it;
+        for (it_min = _conn_stat_map.begin(),
+             it_max = _conn_stat_map.begin(),
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: maybe set it_min/it_max outside(before) the for loop

File: test/apps/iodemo/io_demo.cc [hunk 6]
```diff
@@ -743,12 +797,46 @@ class DemoServer : public P2pDemoCommon {
...
             << " ops, "
-            << "active connections:" << _curr_state.active_conns
-            << ", buffers:" << _data_buffers_pool.allocated()
+            << "active connections:" << _curr_state.active_conns;
+
+        if (_curr_state.active_conns > 0) {
+            assert(!_conn_stat_map.empty());
+
+            log << ", min:" << it_min->second.get_total_count() << "/"
+                << it_min->second.get_total_bytes() / float(UCS_MBYTE)
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: minor: float->double

File: test/apps/iodemo/ucx_wrapper.cc [hunk 7]
```diff
@@ -484,6 +484,12 @@ double UcxContext::get_time() {
     return tv.tv_sec + (tv.tv_usec * 1e-6);
 }
 
+UcxConnection& UcxContext::get_connection(uint64_t id) const {
+    conn_map_t::const_iterator it = _conns.find(id);
+    assert(it != _conns.end());
+    return *(it->second);
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc]: no need ( )

File: test/apps/iodemo/ucx_wrapper.cc [hunk 8]
```diff
@@ -810,6 +819,7 @@ void UcxConnection::established(ucs_status_t status)
     _ucx_status = status;
     _context.remove_connection_inprogress(this);
     if (status == UCS_OK) {
+        print_addresses(false);
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc]: what would be printed when passing false?

File: test/apps/iodemo/ucx_wrapper.h [hunk 9]
```diff
@@ -282,6 +289,8 @@ class UcxConnection {
     uint32_t        _remote_conn_id;
     char            _log_prefix[MAX_LOG_PREFIX_SIZE];
     ucp_ep_h        _ep;
+    std::string     _local_address;
+    std::string     _remote_address;
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.h]: why need to save? can't reuse _log_prefix somehow?

File: test/apps/iodemo/io_demo.cc (hunk 2 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc]: need to define scale and msg size, i believe it's ok for small scale and large msgs

File: test/apps/iodemo/io_demo.cc (hunk 3 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc]: that's the benefit? actually `insert` has one comparison less vs `operator[]` (we expect "insert unique" path) and ability to assert result

File: test/apps/iodemo/io_demo.cc (hunk 4 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:761]: we can but why? my point is similar to `insert` - `find` allows us to assert expected result and erase exact position

File: test/apps/iodemo/ucx_wrapper.cc (hunk 8 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/ucx_wrapper.cc]: actually nothing, I just reused this func to init addresses, need to refactor

**Comment by evgeny-leksikov** [comment]
@yosefe pls take a look

File: test/apps/iodemo/io_demo.cc [hunk 10]
```diff
@@ -609,6 +609,44 @@ class DemoServer : public P2pDemoCommon {
         long    active_conns;
     } state_t;
 
+    class conn_state_t {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: class ConnectionState

File: test/apps/iodemo/io_demo.cc [hunk 11]
```diff
@@ -742,20 +795,56 @@ class DemoServer : public P2pDemoCommon {
...
         memory_pin_stats(&pin_stats);
 
-        LOG << "read:" << _curr_state.read_count - _prev_state.read_count
+        conn_stat_map_t::iterator it_min = _conn_stat_map.begin(),
+                                  it_max = _conn_stat_map.begin();
+        for (conn_stat_map_t::iterator it = _conn_stat_map.begin();
+             it != _conn_stat_map.end(); ++it) {
+
+            long i_val = it->second.get_total_count();
+
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: may remove space line

File: test/apps/iodemo/io_demo.cc [hunk 12]
```diff
@@ -742,20 +795,56 @@ class DemoServer : public P2pDemoCommon {
...
+        log << "read:" << _curr_state.read_count - _prev_state.read_count
             << " ops, "
             << "write:" << _curr_state.write_count - _prev_state.write_count
             << " ops, "
-            << "active connections:" << _curr_state.active_conns
-            << ", buffers:" << _data_buffers_pool.allocated()
+            << "active connections:" << _curr_state.active_conns;
+
+        if (_curr_state.active_conns > 0) {
+            assert(!_conn_stat_map.empty());
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: seems redundant

File: test/apps/iodemo/io_demo.cc [hunk 13]
```diff
@@ -742,20 +795,56 @@ class DemoServer : public P2pDemoCommon {
...
+
+            log << ", min:" << it_min->second.get_total_count() << "/"
+                << it_min->second.get_total_bytes() / double(UCS_MBYTE)
+                << "MB, (" << get_connection(it_min->first).get_peer_name()
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:822]: shall we use get_peer_name in DemoClient as well?

Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:822]: on client side we use `server_name(index)` which goes from `options_t test_opts;`, I think that's OK

**Review by yosefe — APPROVED** [review]

**Pull request merged**